Undeploy on move order for TS units

tovl opened the pull request:
Fixes #13039.
Fixes #16600. 

- Deployed artillery, juggernaut, mobile sensor array and tick tanks now undeploy when given a move order using the force-move modifier.
- Also fixed a minor glitch where EMPed units could not deploy but could still undeploy. They can now do neither.

~~This only applies to units that use the GrantConditionOnDeploy trait, not units that use the Transforms trait, so this does not yet work for MCVs.~~

tovl committed:
Add undeploy on move for GrantConditionOnDeploy and Transforms.

Disallow undeploy when unable to move.

Add undeploy on Move for Transforms.

Make GrantConditionOnDeploy fully compatible with order queues.

Removed unused UndeployForGrantedCondition activity.

Make sure move part is interruptible.

Undeploy on move also applies to aircraft.

Make Transforms fully compatible with order queues.

Punsho commented:
This feature is nice but has also has some negatives, force move is mostly used for crushing infantry and I select my tick tanks with 'select units by type'. What if I want some of my tanks deployed and shooting while others go for crushing? There is no way to select all undeployed tick tanks for this issue to be avoided.

tovl commented:
> What if I want some of my tanks deployed and shooting while others go for crushing?

Ha, I hadn't considered that. Thanks for pointing this out!

> There is no way to select all undeployed tick tanks for this issue to be avoided.

Well, now there is. :wink:

TheChosenEvilOne commented:
IMO This functionality should be enabled by setting something like "UndeployOnMove" to true

tovl commented:
Do you mean setting it per unit type in yaml? Sure, that can be arranged, but I need a bit more context for that:
- What is the use case? 
- What would the alternate behavior need to be?

pchote commented:
I don't like the idea of hardcoding `GrantConditionOnDeploy` in the move activities. I think a better approach here would be to
1. Fix `Mobile` to properly use our pause/disable conventions. The current disabled state should be pause, and then disabled state should remove the cursors etc too.
2. Have the deploy state disable `Mobile`, EMP etc pause it.
3. Let `GrantConditionOnDeploy` handle the move cursors, and then issue the move order itself queued after the undeploy.
4. Do the same thing in `Transforms`, using a new `MoveLocationInit` to pass the requested move target to th enew actor.

tovl commented:
I'm busy rewriting this now. However, I'm stuck at step 4. I can pass the target location to the new actor, but how do I make it actually perform a move activity?

pchote commented:
If you didn't already find it, use `ITransformActorInitModifier` to insert the target location into the initializer dictionary for the new actor.  Then in the `Mobile` and `Aircraft` constructors pull it out and store it in a field. Pull the contents of the `if (order.OrderString == "Move")` blocks in `ResolveOrder` out into their own methods, and then call this with the stashed target location from the `Created` method - note that you can't do this directly from the constructor because other traits on the actor may not be created yet.

tovl commented:
Thanks! Yeah, I tried starting the activity from the constructor, but that didn't work. It sounds like the `Created` method is what I was missing.

tovl commented:
Reimplemented according to the approach suggested by pchote. Additionaly:
- Undeploy on move order now also works with the `Transforms` trait
- Added a yaml parameter to `GrantConditionOnDeploy` and `Transforms` to undeploy-on-move either:
  - Always with or without force-move
  - Only on force-move
  - Never
- Deploying/undeploying with `GrantConditionOnDeploy` is now completely compatible with order queues. Everything can be queued before or after a deploy/undeploy order including other deploy/undeploy orders. Furthermore, queued orders are not ignored when in the middle of deploying/undeploying (unqueued orders are still ignored). In order to do so I have merged `UndeployForGrantedCondition` into `DeployForGrantedCondition`. 
- Fixed a minor visual glitch on the Nod Artillery where the barrel temporarily disappeared when deploying.
- In case my first comment was a bit cryptic: Deployed and undeployed units are now selected separately when using the "W" hotkey or when double clicking on a unit.

The code does not yet work for aircraft because the OrderTargeters rely on knowing `LocomotorInfo.MoveIntoShroud` and `LocomotorInfo.MovementCostForCell`. I guess this should go into an interface? However, I'm not sure it's a good idea bothering with this now, when the whole Aircraft trait is in the middle of revision and there are no flying units who need this in the default mods anyway.

Punsho commented:
Will it work on construction yards? It would be really nice if you could click on a con yard, force move it, queue a deploy. It would require less attention and would be more efficient overall

tovl commented:
>  It would be really nice if you could click on a con yard, force move it, queue a deploy.

That was actually what I was thinking as well. Unfortunately, MCVs work with `Transforms`, not `GrantConditionOnDeploy`, so this is not yet possible. Making this possible for `Transforms` would be much more complicated, I fear. But, seeing that there are more people interested in this than just myself, I'll see how far I can get.

tovl commented:
This should now fully work for aircraft as well, although I cannot test it since there are no relevant units in the default mods.

pchote commented:
I'd prefer if the "separate selection criteria" commit could be handled in a separate PR, and then implemented using conditions so that it could fix #15370.

pchote commented:
Can you then please squash/rework the remaining commits so the history only contains the final implementation?

pchote commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
-    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
```
Please drop this and the formatting changes below

tovl replied:
Ugh, this seems to be something that mono does on it's own sometimes. Very annoying.

tovl replied:
Fixed.

tovl commented:
I guess that means `Selectable` would need to be moved to OpenRA.Mods.Common?

pchote commented:
The idea from #4783 was to move it onto `Mods.Common` and merge it with `SelectionDecorations` as a prereq for implementing the isometric selection boxes for TS/RA2 buildings.

tovl commented:
- `Transforms` now carries over the complete activity queue to the newly created actor, which makes it fully compatible with queued orders. The behavior of MCVs should now be identical to the behavior of other deployable units introduced earlier.
- The changes to selection criteria are removed and will soon be posted to a new PR.
- Squashed some commits.

pchote commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -990,4 +990,4 @@
   </Target>
-</Project>
\ No newline at end of file
+</Project>
```
Style nit: all our editors fight over formatting, so our convention is to only commit the essential project file changes in each PR.

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Scripting/Properties/TransformProperties.cs:
```diff
@@ -30,7 +30,7 @@ public TransformProperties(ScriptContext context, Actor self)
 		}
 	}
-}
+}
```
This can be unstaged too

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -171,6 +171,7 @@ public class Aircraft : ITick, ISync, IFacing, IPositionable, IMove, IIssueOrder
 
 		public readonly AircraftInfo Info;
 		readonly Actor self;
+		readonly Target InitTarget;
```
Naming nit `initialTarget`

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -847,6 +838,24 @@ public void ResolveOrder(Actor self, Order order)
 			}
 		}
 
+		void DoMove(Target target, bool queued)
```
Naming nit: aim for something like `void MoveInner(Actor self, Target target, bool queued)`

tovl replied:
This is no longer relevant in the current version.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -847,6 +838,24 @@ public void ResolveOrder(Actor self, Order order)
 
+		void DoMove(Target target, bool queued)
+		{
+			var cell = self.World.Map.Clamp(this.self.World.Map.CellContaining(target.CenterPosition));
```
omit `this`

tovl replied:
No longer relevant in the current version.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs:
```diff
@@ -54,6 +54,9 @@ public class GrantConditionOnDeployInfo : PausableConditionalTraitInfo
 		public readonly bool SkipMakeAnimation = false;
 
+		[Desc("Can this actor be ordered to move when deployed? [always, never, force-only]")]
+		public readonly string UndeployOnMove = "always";
```
Always prefer enums to magic strings - the `FieldLoader` can automatically turn `Always` in the yaml into `UndeployType.Always` (or whatever we want to call the enum and values).

tovl replied:
Turned `UndeployOnMove` into an enum.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs:
```diff
@@ -298,36 +301,37 @@ class MoveDeployOrderTargeter : IOrderTargeter
-				mobile = self.TraitOrDefault<Mobile>().Info;
+				mobile = actorInfo.TraitInfoOrDefault<MobileInfo>();
 				locomotorInfo = mobile.LocomotorInfo;
-				rejectMove = !self.AcceptsOrder("Move");
```
We need to be careful about this, there are cases in missions where we want actors to be able to move in response to map scripts, have their weapons controlled by the player, but reject any player given move orders. This was implemented via the `RejectsOrders` trait and this check.

tovl replied:
`rejectMove` is put back in.

pchote commented on OpenRA.Mods.Common/Traits/Transforms.cs:
```diff
@@ -124,20 +126,88 @@ public void DeployTransform(bool queued)
 			if (!queued)
 				self.CancelActivity();
 
+
```
Stray whitespace

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs:
```diff
@@ -299,7 +300,8 @@ public bool TargetOverridesSelection(TargetModifiers modifiers)
 				this.unit = unit;
-				info = actorInfo.TraitInfoOrDefault<MobileInfo>();
+				mobileInfo = actorInfo.TraitInfoOrDefault<MobileInfo>();
+				aircraftInfo = actorInfo.TraitInfoOrDefault<AircraftInfo>();
```
Thinking out loud: we could add `MoveIntoShroud` and `CanEnterCell` (is this what the `MovementCostForCell` check was doing?) on the `IMoveInfo` interface, and then remove the explicit references to `MobileInfo` and `AircraftInfo` here and in Transforms.

tovl replied:
I think that would be an elegant solution. However, `CanEnterCell` is defined slightly differently for `AircraftInfo`. It checks whether the aircraft can land in the cell not whether it can move in the cell.

tovl replied:
I've implemented this using a new method `CanMoveInCell` to `IMoveInfo` that returns `CanEnterCell` in `Mobile` and always returns true in `Aircraft`.

pchote reviewed (commented):
Some more detailed thoughts, and it looks like this will want to depend on #16246 too.

tovl committed:
Add CanMoveInCell and GetMoveIntoShroud to IMoveInfo.

tovl commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -125,7 +125,7 @@ IEnumerable<EditorActorOption> IEditorActorOptions.ActorOptions(ActorInfo ai, Wo
 	}
 
-	public class Mobile : ConditionalTrait<MobileInfo>, INotifyCreated, IIssueOrder, IResolveOrder, IOrderVoice, IPositionable, IMove,
+	public class Mobile : PausableConditionalTrait<MobileInfo>, INotifyCreated, IIssueOrder, IResolveOrder, IOrderVoice, IPositionable, IMove,
```
Split off and amended; See #16262.

tovl commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -760,14 +760,17 @@ public bool CanTarget(Actor self, Target target, List<Actor> othersAtTarget, ref
 				if (rejectMove || target.Type != TargetType.Terrain)
 					return false;
 
+				if (mobile.IsTraitDisabled)
```
Fixed in #16262.

tovl commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -706,7 +706,7 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 				if (!order.Queued)
 					self.CancelActivity();
-
+					
```
Fixed.

tovl commented on OpenRA.Mods.Common/Traits/Transforms.cs:
```diff
@@ -124,20 +126,88 @@ public void DeployTransform(bool queued)
+		}
+	}
+
+	public class MoveLocationInit : IActorInit<Target>
```
This is no longer relevant in the current version.

tovl commented:
Update:
- Refactored the code somewhat and addressed the above concerns as well as making sure the code works well with #16246 and #16262.
- Added new methods to `IMove` interface.
- Squashed commits further to prevent redundant review work in the future (sorry about that).
- `Transforms` is no longer disabled by `build-incomplete.` This is to ensure deploy orders are still registered when transitioning and make the trait more consistent with `DeployforGrantedCondition`. 
- In order to prevent the build animation from being interruptible I have used a `Wait` activity. I am not entirely happy with this, because that means the deployment length is not dependent on the animation length. (Of course you can interpret this as an advantage in some respects...) I'm not sure how else to do this without scope-creeping again. One way to handle this could be to make the build animation itself into an Activity that is not interruptible, which might also be useful for handling  #16250 as well and might open up lots of interesting possibilities. However, like I said, this would be way out of scope and I'd rather use a quick fix here.

pchote commented:
#16246 has been merged, so this needs a rebase now.

tovl commented:
Rebased.

abcdefg30 commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -421,7 +421,11 @@ public interface ITargetablePositions
-	public interface IMoveInfo : ITraitInfoInterface { }
+	public interface IMoveInfo : ITraitInfoInterface
+	{
+		bool GetMoveIntoShroud();
```
Imho this would make more sense as `CanMoveIntoShroud`. Any other opinions?

tovl replied:
Makes sense to me. Renamed.

abcdefg30 reviewed (dismissed):
Seems to work nice ingame. Needs another rebase, though.

pchote commented:
I'd like to take another close look at this before merging, as I suspect I may have some more suggestions on the plumbing. Lets get the other PRs out of the way first, though.

tovl commented on mods/ts/rules/gdi-vehicles.yaml:
```diff
@@ -345,12 +345,15 @@ JUGG:
 	WithMoveAnimation:
 		MoveSequence: walk
 	GrantConditionOnDeploy:
+        PauseOnCondition: empdisable || being-captured
```
Fixed.

tovl commented:
Rebased.

obrakmann commented on mods/ts/rules/gdi-vehicles.yaml:
```diff
@@ -345,12 +345,15 @@ JUGG:
 		AllowedTerrainTypes: Clear, Road, DirtRoad, Rough
 		DeploySounds: place2.aud
 		UndeploySounds: clicky1.aud
+		UndeployOnMove: ForceOnly
```
I'd set this the other way around. Always for units and ForceOnly for the conyards, on all mods. This is otherwise too easy to trigger and would probably lead to frustrations when playing.

dragunoff replied:
I think it should definitely be `ForceOnly` for conyards. I'd even argue that `ForceOnly` should be the default because of the same reason - frustration and anger when you undeploy by accident (could happen if you misclick when shift queuing attack orders for example).

tovl replied:
I don't remember it ever being a problem with conyards in the original RA2 and why should it? Undeploying is basically the only reason to even select a conyard and then issue and order in the first place. I do remember it being sometimes annoying with artillery in TS because you want to order an attack but misclick (but then again isn't that an issue with all offensive units?).

Regardless, I'm willing to go with the flow here. However, since this is just a simple yaml switch that can be changed with minimal effort at any time in the future, I'll propose instead to leave it at `Always` for all units initially and use future playtests to empirically determine whether and where this is problematic.

obrakmann commented on OpenRA.Mods.Common/Activities/DeployForGrantedCondition.cs:
```diff
@@ -18,44 +18,60 @@ namespace OpenRA.Mods.Common.Activities
+				if (deploy.DeployState == DeployState.Undeployed)
+				{
+					if (orderedMove && move != null)
+						QueueChild(self, move.MoveTo(cell, 8), true);
```
Hmm, do I get this right? When we're taking this branch, no deploy will ever actually be executed, correct? So this is basically just wrapping around a regular move order for undeployed units? Is that intentional?

tovl replied:
Yes, this is intentional. All move orders are intercepted by `GrantConditionOnDeploy` and interpreted on the fly by `DeployForGrantedCondition` as either regular moves or undeploy-moves depending on the state at the moment the activity is executed. This ensures that move orders are always interpreted correctly no matter how convoluted the activity-queue gets.

obrakmann commented on OpenRA.Mods.Common/Activities/DeployForGrantedCondition.cs:
```diff
@@ -64,4 +80,40 @@ public override Activity Tick(Actor self)
+
+			if (!initiated)
+			{
+				if (towardDeploy)
```
Can't you just look at the `DeployState` and figure out from that what to do?

tovl replied:
Hmm, it seems that way, yes. I can't now remember what the reasoning behind that was.

pchote replied:
Is this comment still valid?

tovl replied:
Yep. Fixed it now.

obrakmann commented on OpenRA.Mods.Common/Activities/Transform.cs:
```diff
@@ -28,61 +28,94 @@ public class Transform : Activity
+					hasMoved = true;
+					QueueChild(self, move.MoveTo(cell, 8), true);
+					return this;
+				}
```
Same as the previous comment: this is just wrapping around a regular move order for undeployed actors, right?

tovl replied:
Yes, see above.

obrakmann reviewed (commented):
Seems to work as promised in-game. Just a couple of questions/remarks, though.

pchote commented:
Deployed units show the "enter tunnel" cursor and give voice feedback when ordered, but don't actually undeploy or move.  I wonder if we can avoid this and probably other issues by adapting the approach here to keep Mobile in charge of the cursors and orders, but letting `Transforms` and `GrantConditionOnDeploy` hook into it via a new interface.

tovl commented:
>I wonder if we can avoid this and probably other issues by adapting the approach here to keep Mobile/Aircraft in charge of the cursors and orders, and introduce a new interface implemented by Transforms and GrantConditionOnDeploy that is used to disable the move cursor visibility (e.g. when ForceOnly and the modifier isn't held) and to fetch an activity to queue immediately before the move.

> I know this rolls back my original suggestions a bit, but it should still avoid the coupling I wanted to avoid, hopefully without as many side-effects.

I don't think flip-flopping on the basic approach is productive. I'd rather just work out the kinks in the current approach instead.

pchote commented:
My concern is that the current approach won't be able to handle attack-move, entering tunnels / transports, or any other orders that move but aren't "Move".  Having all of these hardcoded in the deploy activities would mean that mods couldn't add their own.

tovl commented:
That would basically mean throwing away everything and starting from scratch (again). It also probably won't work for `Transforms`. That leaves me with little motivation to continue this.

pchote commented:
If there is another way to solve the non-move activities then lets consider it, otherwise I dont mind taking this over.

tovl commented:
Keep in mind that a deployed conyard doesn't have an `IMove` trait, so I'm afraid that keeping the `IMove` trait in charge would just be a dead end.

Thinking about it some more, attack-move and enter basically work the same way as the current approach here (wrapping themselves around the move activities). The problem exists there as well: You can't attack-move into a tunnel or into a transport for example. The idea of wrapper orders makes sense, but what we need is a way to have wrappers wrap around each other (sort of like decorator functions). If we can solve it this way, we also solve it for all other interaction between "move but not move" orders.

pchote commented:
One potential solution for the construction yard would be to create a `TransformedActorIsMobile` trait that handles the bits that are needed for the cursors and orders without directly implementing `IMove`

Attack-moving into a transport isn't something we expose for the player, so i'm not sure that needs to be a blocker. This is an interesting idea, though.

tovl commented:
What are actually the blockers for dropping `Transforms` entirely and just let MCVs use `GrantConditionOnDeploy` instead?

tovl committed:
Let GrantConditionOnDeploy and Transforms Wrap IMove.

pchote commented:
The biggest issue is dealing with `Building` &lrarr; `Mobile` and the associated cell reservation and location changes.

tovl commented:
I've solved the problem with the "move but not `Move"` orders for `GrantConditionOnDeploy` and refactored `Transforms` to use the same pattern. However, I'm still not sure how to actually solve it for `Transforms`. I'm not really sure what a potential `TransformedActorIsMobile` trait should do here and if adding another single-use trait is the best solution anyway.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -491,29 +498,49 @@ public void RemoveInfluence()
 
 		public Activity MoveTo(CPos cell, int nearEnough)
 		{
+			var moveWrapper = self.TraitOrDefault<IWrapMove>();
```
We can reduce the duplication and trait queries by doing something like:
```csharp
// Lives up with the other trait references and assigned from Created using self.TraitsImplementing<IWrapMove>().ToArray();
IWrapMove[] moveWrappers;

Activity WrapMove(Func<Actor, Activity> inner)
{
	var moveWrapper = moveWrappers.FirstOrDefault(Exts.IsTraitEnabled);
	if (moveWrapper != null)
		return moveWrapper.WrapMove(inner);

	return inner(self);
}

public Activity MoveTo(CPos cell, int nearEnough)
{
	return WrapMove(self => new Move(self, cell, WDist.FromCells(nearEnough)));
}

// ... and so on for the others
```

... but see my question below, as this may change if we want to change the way the wrapping works.

tovl replied:
Deduplicated the code as suggested.

pchote commented on OpenRA.Mods.Common/Activities/DeployForGrantedCondition.cs:
```diff
@@ -19,48 +20,86 @@ public class DeployForGrantedCondition : Activity
 	{
 		readonly GrantConditionOnDeploy deploy;
 		readonly bool canTurn;
+		readonly Func<Activity> moveInner;
```
Can you elaborate on the practical benefits of `Func<Activity>`s vs evaluating the activity up-front and either storing it here as an `Activity` or, I hope better, queuing it as `NextActivity`?

tovl replied:
The idea was for `Transforms` to evaluate the activity as late as possible to prevent trying to look up traits from a dead actor and causing crashes. This still didn't work out right though.

I like the idea of queuing the activity instead and have now used other means to prevent dead actor trait look-ups.

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -420,6 +420,12 @@ public interface IMove
+	public interface IWrapMove
+	{
+		Activity WrapMove(Func<Actor, Activity> moveInner);
+		bool CanWrapOrder(TargetModifiers modifiers);
```
`CanWrapMoveOrder` for consistency please.

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Traits/EntersTunnels.cs:
```diff
@@ -92,6 +92,11 @@ public override bool CanTargetActor(Actor self, Actor target, TargetModifiers mo
 
+				var wrapMove = self.TraitOrDefault<IWrapMove>();
+				var move = self.TraitOrDefault<Mobile>();
+				if (move != null && move.IsTraitDisabled && (wrapMove == null || !wrapMove.CanWrapOrder(modifiers)))
```
Why do tunnels deserve a special case, but not for example entering a transport?

tovl replied:
It doesn't make much sense for `AttackMove` to enforce the required modifier in addition to already needing to select "A", but for the other cases this is an oversight. Added the same check in `EnterAlliedActorTargeter`

pchote reviewed (commented):
I'm still wrapping my head around this, but have a couple of initial questions/comments:

pchote commented:
>  However, I'm still not sure how to actually solve it for `Transforms`

I checked vanilla TS, and it cops out by only supporting the move cursor on standard clear terrain. You get a move-blocked cursor when mousing over tunnels and even tiberium.

tovl commented:
I've finally managed to get it working for `Transforms` in combination with a dummy `Mobile` trait as suggested. The only thing that still doesn't work is ordering a construction yard to a repair depot, but that has more to do with the `Repairable` trait than with `Transforms` (specifically `AfterReachActivities` needs to be a proper activity). With all the upcoming Resupply/Repair refactoring I don't feel particularly inclined to fix that here, so I've just disabled it for ConYards for now.

> I checked vanilla TS, and it cops out by only supporting the move cursor on standard clear terrain. You get a move-blocked cursor when mousing over tunnels and even tiberium.

I can understand why; Having actors switch out like that gets pretty head-ache inducing :stuck_out_tongue:

pchote commented:
Adding this to the release milestone, but I expect to wait until after the aircraft activities and #16481 are merged before looking at this in detail again.

tovl commented:
Rebased.

pchote commented:
Transferring orders created for one actor over to another actor is asking for a lot of trouble:
* Activities are very likely to regress in the future when someone changes them without realizing the non-conventional use here
* It forces us to remove performance optimizations (e.g. caching trait references) from any activities that may be transferred
* It breaks potentially many other assumptions, which will lead to difficult to debug issues

At the risk of falling into the "everything is a &lt;X&gt;" trap, I do think our newly introduced activity interface concept provides a clean solution that doesn't require another total rewrite here:

* Define an `IActivityTranferrableToTransformedActor` interface with a method `Order IssueOrderForTransformedActor(Actor self, Actor newActor)`.
* Implement the interface on `Move`, `MoveAdjacentTo`, etc.
* Replace the `init.Add(new ActivityInit(self.CurrentActivity))`  with a loop over `ActivitiesImplementing<IActivityTranferrableToTransformedActor>` calling `a.ResolveOrder()` with each order on the newly created actor.
* Undo the changes to `Enter`, `MoveWithinRange`, etc.

This will mean introducing a couple of new order types to be resolved by `Mobile` (MoveAdjacentTo, MoveWithinRange), but I feel this is going to be well worth the trade off in terms of robustness and avoiding perf regressions.

A nice side effect is that we or third party modders could extend the system with additional traits (e.g. `TransformToPassenger`) without having to make big changes to core code.

pchote commented:
It may be possible to implement this in a simpler way by adding a single `IssueOrderAfterTransform` activity that the `TransformTo*` trait(s) queue, without relying on an interface.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs:
```diff
@@ -57,17 +58,23 @@ public class GrantConditionOnDeployInfo : PausableConditionalTraitInfo
 		public readonly string Voice = "Action";
 
+		[Desc("Can this actor be ordered to move when deployed? [Always, ForceOnly, Never]")]
+		public readonly UndeployOnMoveType UndeployOnMove = UndeployOnMoveType.Always;
```
The default behavior for `GrantConditionOnDeploy` is to have nothing to do with movement, so the default here really should be `UndeployOnMoveType.Never` to match. Mods must explicitly opt-in to having the deploy affect movement, so it it logical for them to opt in to having movement affect the deploy at the same time.

This otherwise breaks usecases that have nothing to do with movement - e.g. deploy to activate gap generation or cloak.

pchote commented:
I can take this over. This is real close to being mergable, so it would be a shame to leave it at the (hopefully) last hurdle.

pchote commented:
16e0d5517611a4402e62b643e786c5a6c3616aa5 implements my suggestion above, which seems to work well enough.

I'll open a new PR once I have a chance to go through in more detail and make sure I understand the rest of the changes, apply any other comments I might have, and maybe split the last commit into a few smaller ones.
